crates/oxc_allocator/src/clone_in.rs

```diff
@@ -0,0 +1,89 @@
+use crate::{Allocator, Box, Vec};
+
+/// A trait to explicitly clone an object into an arena allocator.
+///
+/// As a convention `Cloned` associated type should always be the same as `Self`,
+/// It'd only differ in the lifetime, Here's an example:
+///
+/// ```
+/// impl<'old_alloc, 'new_alloc> CloneIn<'new_alloc> for Struct<'old_alloc> {
+///     type Cloned = Struct<'new_alloc>;
+///     fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned {
+///         Struct { a: self.a.clone_in(alloc), b: self.b.clone_in(alloc) }
+///     }
+/// }
+/// ```
+///
+/// Implementations of this trait on non-allocated items usually short-circuit to `Clone::clone`;
+/// However, it **isn't** guaranteed.
+///
+pub trait CloneIn<'new_alloc>: Sized {
+    type Cloned;
+
+    fn clone_in(&self, alloc: &'new_alloc Allocator) -> Self::Cloned;
+}
+
+impl<'a, T, C> CloneIn<'a> for Option<T>
+where
+    T: CloneIn<'a, Cloned = C>,
+{
+    type Cloned = Option<C>;
+    fn clone_in(&self, alloc: &'a Allocator) -> Self::Cloned {
+        self.as_ref().map(|it| it.clone_in(alloc))
+    }
+}
```

Comment:
Can we use `'new_alloc` here instead of `'a`? I know there's only 1 lifetime on this impl, but as you've pointed out, this trait is somewhat confusing, so I think it's better to be explicit.

Ditto for `Cell` impl, and `impl_clone_in!`.

Reply:
Higher up the stack I've used `'alloc` for implementations with only one allocator lifetime. Do you think it is sufficient or should I go ahead and refactor them all?

https://github.com/oxc-project/oxc/blob/383afeaff1bcd3bd4ef4170d5a0823180d79a928/crates/oxc_ast/src/generated/derive_clone_in.rs#L14

Reply:
Personally I think it's preferable to use `'new_alloc` here so it matches the rest of the impls in this file. But I can see argument for `'alloc` for consistency. Up to you.